spellmonitor.lic

```diff
@@ -43,139 +40,121 @@ class DRSpells
-                     when /(.+)  \(OM\)/
-                       DRSpells.active_spells[Regexp.last_match(1)] = 1000 # is there a better solution?
-                       DRSpells.refresh_data[Regexp.last_match(1)] = true
-                     when /.*orbiting sliver.*/
-                       DRSpells.slivers = true
-                     when /(.+)  \((\d+)%\)/
-                       DRSpells.active_spells[Regexp.last_match(1)] = Regexp.last_match(2).to_i
-                       DRSpells.refresh_data[Regexp.last_match(1)] = true
-                     when /^*$/
-                       DRSpells.active_spells[Regexp.last_match(1)] = 1000 # is there a better solution?
-                       DRSpells.refresh_data[Regexp.last_match(1)] = true
-                     end
-                   end
-
-                   server_string
-                 end
-               end
+# Some spells may last for an unknown duration,
+# such as cyclic spells that last as long as
+# the caster can harness mana for it.
+# Or, barbarian abilities when the character
+# doesn't have Power Monger mastery to see true
+# durations but only vague guestimates.
+# In those situations, we set use this value.
+UNKNOWN_DURATION = 1000
+
+# This method receives a line of data from DragonRealms
+# and determines if it refers to active spells on the character.
+# It keeps the global class DRSpells updated with such data
+# that then other scripts can take advantage of.
+#
+# The XML stream for DragonRealms is whack at best.
+#
+# Examples of the data we're looking for:
+#     <clearStream id="percWindow"/>
+#     <pushStream id="percWindow"/> Wildfire (thoroughly inflamed)
+#     <popStream/><pushStream id="percWindow"/> Avalanche (thoroughly inflamed)
+#     <popStream/><pushStream id="percWindow"/>Righteous Wrath  (8 roisaen)
+#     Rage of the Clans  (4 roisaen)
+#     <popStream/><prompt time="1609379003">&gt;</prompt>
+#
+# <clearStream> indicates that the following tags are the latest active spells snapshot.
+#
+# <pushStream id="percWindow"> indicates a spell or spells that are active and their duration.
+#
+# <popStream/> indicates the end of content for the prior listed spell or spells for the <pushStream> tag.
+#
+# One or more spells may be listed between a <pushStream/> <popStream/> pair,
+# but only one spell and its duration are ever listed per line.
+#
+spell_action = proc do |server_string|
+  # New information about active spells, reset our cached data.
+  if server_string =~ %r{<clearStream id="percWindow"/>}
+    DRSpells.slivers = false
+    DRSpells.refresh_data.each_key { |k| DRSpells.refresh_data[k] = false }
+    begin
+      Thread.new do
+        sleep 0.5
+        DRSpells.refresh_data.each do |key, state|
+          unless state
+            DRSpells.active_spells.delete(key)
+            DRSpells.refresh_data.delete(key)
+          end
+        end
+      end
+    rescue StandardError => e
+      echo(e)
+      echo('Error in spell monitor')
+    end
```

Comment:
I don't know why a thread is used here to remove keys from the map, this is original logic